Add class to generate and validate tokens for site authentication

fjorgemota opened the pull request:
Fixes 413-gh-Automattic/wpjobmanager.com
Sibling PR to 445-gh-Automattic/wpjobmanager.com

### Changes proposed in this Pull Request

- Add class responsible for:
    -  creating and validating tokens on WPJM, associated with a post OR user
    -  validate the tokens created 

### Testing instructions
 
Run the unit tests, and experiment creating tokens using WP-CLI Shell:

```php
$instance = WP_Job_Manager_Com_Auth_Token::instance(); // To get the instance of the class
$token = $instance->generate( 'user', 1 ); // Create a token associated with the User ID 1
// Maybe await a minute using sleep(60)?
var_dump($instance->validate( 'user', 1, $token));
// Make the same test with a valid post instead of user
```

fjorgemota committed:
Add methods to encode and persist the token for site authentication

fjorgemota committed:
Add methods to validate if a token is valid or not, and delete them appropriately from the database

fjorgemota committed:
Add documentation to the constants in WP_Job_Manager_Com_Auth_Token

fjorgemota committed:
Fix spacing between methods in WP_Job_Manager_Com_Auth_Token

fjorgemota committed:
Make is_valid method private in the class to manage tokens for WPJMCOM

fjorgemota committed:
Include class for creating Site Auth tokens in WP Job Manager Helper

fjorgemota committed:
Add singleton method to WP_Job_Manager_Com_Auth_Token class

fjorgemota committed:
Fix typo on accepted object type for user on the class responsible for creating tokens

fjorgemota committed:
Fix issue when validate is called with an invalid object ID

fjorgemota committed:
Add unit test for the WP_Job_Manager_Com_Auth_Token class

fjorgemota commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,168 @@
+ * @group helper
+ * @group helper-base
+ */
+class WP_Test_WP_Job_Manager_Com_Auth_Token extends WPJM_BaseTest {
```
I'm not sure what the best way to test expired tokens here..if anyone has any suggestions, please feel free to mention it. :)

renatho replied:
IMO, the way you implemented it by updating the meta is good enough.

fjorgemota replied:
Yeah, I implemented that this morning. Thanks for the feedback!

fjorgemota commented on includes/helper/class-wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,177 @@
+	 *
+	 * @return string|WP_Error The random token generated, or a WP_Error if the token could not be generated.
+	 */
+	private function generate_new_token() {
```
This method will generate a string encoded as base64, but without some characters to make it URL-Safe. Because of this, it generates a string with varied size, with up to 64 characters.

fjorgemota committed:
Test logic to remove site auth tokens that were used once

fjorgemota committed:
Use base64 to encode token, and use 48 random bytes to compute token

fjorgemota committed:
Add test to verify if validate return false when token is expired

fjorgemota committed:
Add methods to verify if validate deletes expired tokens automatically

fjorgemota committed:
Fix singleton type on WP_Job_Manager_Com_Auth_Token

renatho commented on includes/helper/class-wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,179 @@
+			return false;
+		}
+		foreach ( $metadatas as $metadata ) {
+			if ( ! $this->is_valid( $metadata ) ) {
```
WDYT to rename it to `is_valid_format` or something similar? It seems the main validation method.

fjorgemota replied:
Fair, renamed in 5f0ea4b519579ce82e49ddb2c9abb50dee44cfb4. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+ */
+class WP_Test_WP_Job_Manager_Com_Auth_Token extends WPJM_BaseTest {
+
+	public function testInstance_WhenCalled_ReturnSameInstance() {
```
Not sure if I misunderstood something, but aren't these 2 tests with the name exchanged? (`testInstance_WhenCalled_ReturnSameInstance` and `testInstance_WhenCalled_ReturnCorrectType`)

fjorgemota replied:
One tests if the instance returned is the same, another tests if it returns the correct type.

IMHO it is kinda a boilerplate test, as it is testing..the singleton (which isn't technically even needed by this class), but oh well

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+
+		// Act.
+		$result = $instance->generate( 'user', "test");
```
```suggestion
		$result = $instance->generate( 'user', 'invalid_id' );
```

I suggested a format fix and I took the opportunity to just suggest another string to make the test more explicit.

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		$user = $this->factory->user->create_and_get();
+
+		// Act.
+		$result = $instance->generate( 'user', $user->ID);
```
```suggestion
		$result = $instance->generate( 'user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		$post = $this->factory->post->create_and_get();
+
+		// Act.
+		$result = $instance->generate( 'post', $post->ID);
```
```suggestion
		$result = $instance->generate( 'post', $post->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$user = $this->factory->user->create_and_get();
+		$token = $instance->generate('user', $user->ID);
```
```suggestion
		$token = $instance->generate('user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$post = $this->factory->post->create_and_get();
+		$token = $instance->generate('post', $post->ID);
```
```suggestion
		$token = $instance->generate( 'post', $post->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$user = $this->factory->user->create_and_get();
+		$token = $instance->generate('user', $user->ID);
```
```suggestion
		$token = $instance->generate( 'user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$post = $this->factory->post->create_and_get();
+		$token = $instance->generate('post', $post->ID);
```
```suggestion
		$token = $instance->generate( 'post', $post->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$user = $this->factory->user->create_and_get();
+		$instance->generate('user', $user->ID);
```
```suggestion
		$instance->generate( 'user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$post = $this->factory->post->create_and_get();
+		$instance->generate('post', $post->ID);
```
```suggestion
		$instance->generate( 'post', $post->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$user = $this->factory->user->create_and_get();
+		$token = $instance->generate('user', $user->ID);
```
```suggestion
		$token = $instance->generate( 'user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		$this->assertFalse( $result );
+	}
+
+	private function expire_tokens ( $object_type, $object_id ) {
```
```suggestion
	private function expire_tokens( $object_type, $object_id ) {
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$post = $this->factory->post->create_and_get();
+		$token = $instance->generate('post', $post->ID);
```
```suggestion
		$token = $instance->generate( 'post', $post->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho commented on tests/php/tests/includes/helper/test_class.wp-job-manager-com-auth-token.php:
```diff
@@ -0,0 +1,263 @@
+		// Arrange.
+		$instance = WP_Job_Manager_Com_Auth_Token::instance();
+		$user = $this->factory->user->create_and_get();
+		$token = $instance->generate('user', $user->ID);
```
```suggestion
		$token = $instance->generate( 'user', $user->ID );
```

fjorgemota replied:
Fixed in 3e75085c236778ceb0fe8e4f055aca2b6a1323e3. Thanks!

renatho reviewed (commented):
Sorry that you will see a lot of comments, but almost all of them are related to PHPCS, which seems to not be working on GH Actions.

I liked the code. It seems very organized and clear.
Just to understand, I'd like just to confirm your idea about user and post for the Promoted Jobs project. Are you planning to generate a token associated with the job? Or with the user?

fjorgemota committed:
Fix style of the code in the test class

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

fjorgemota committed:
Rename is_valid to is_valid_format in WP_Job_Manager_Com_Auth_Token

fjorgemota committed:
Merge remote-tracking branch 'origin/add/site-auth-token' into add/site-auth-token

fjorgemota commented:
>  Just to understand, I'd like just to confirm your idea about user and post for the Promoted Jobs project. Are you planning to generate a token associated with the job? Or with the user?

The idea of supporting `post` is to allow us to easily work with this issue: 433-gh-Automattic/wpjobmanager.com - and based on two assumptions:

- We will not use a `WP_User` in WPJMCOM to represent each site
- Most of this class will be shared with WPJMCOM, just with a few changes, like namespace, name of the class, etc.

Hope this answers your question. :)

aaronfc commented:
> Most of this class will be shared with WPJMCOM, just with a few changes, like namespace, name of the class, etc.

I am not sure I understand what is the purpose of having this code in WPJMCOM. AFAIK we will generate and validate the token on WPJM site only, right? On WPJMCOM I would expect only to "receive" a token and do a query to some public endpoint on the WPJM to assert that the result is "valid". Maybe I am missing something else?

aaronfc reviewed (commented):
> Most of this class will be shared with WPJMCOM, just with a few changes, like namespace, name of the class, etc.

I am not sure I understand what is the purpose of having this code in WPJMCOM. AFAIK we will generate and validate the token on WPJM site only, right? On WPJMCOM I would expect only to "receive" a token and do a query to some public endpoint on the WPJM to assert that the result is "valid". Maybe I am missing something else?

A part from that, I am also a little confused about how we will use this for the basic promote job flow. Would we use the 'user' type there or also the 'post' for the Job we want to promote? It's a bit confusing to me that this is "site authentication" but there is no way to authenticate the "site" as a whole (only users or posts from the site).

fjorgemota commented:
Hey there!

> I am not sure I understand what is the purpose of having this code in WPJMCOM. AFAIK we will generate and validate the token on WPJM site only, right? On WPJMCOM I would expect only to "receive" a token and do a query to some public endpoint on the WPJM to assert that the result is "valid". Maybe I am missing something else?

No, we will need this on WPJMCOM too, to work with this issue:  433-gh-Automattic/wpjobmanager.com, mentioned above, which will implement the reverse flow with the purpose of protecting #2442:

1. WPJMCOM receives the confirmation that the job was published by JobTarget
2. WPJMCOM generates a token (using this same system defined here), and associates it with the Promoted Site CPT (hence why this also supports `post`, more on this later);
3. It contacts the WPJM Site, more specifically the API defined in #2442, to update the job status, passing the token generated above, and probably the Promoted Site ID too (if the WPJM Site doesn't know it already)
4. WPJM Site validates the request is indeed coming from WPJMCOM (otherwise, anyone can update the job status, which would be annoying)
5. WPJMCOM verifies if the token is valid or not
6. WPJM Site proceeds with updating or not the job status

Hope this answer makes it clear why we need to have a very similar code on WPJMCOM, and also why we're supporting `post`.

> A part from that, I am also a little confused about how we will use this for the basic promote job flow. Would we use the 'user' type there or also the 'post' for the Job we want to promote?

The way I see it on the WPJM side we can either associate with the user OR with the post for the job we want to promote. . IMHO, it would make more sense to associate it with the user (e.g. so a token generated for me can't be validated for another user of the website, although that won't make a real difference in the end, to be honest), but keep in mind that, we'll need to share the ID of the thing with WPJMCOM. Are we associating the token with the user? We'll need to share the User ID. Are we associating the token with the listing being promoted? We'll need to share that listing's ID.

>  It's a bit confusing to me that this is "site authentication" but there is no way to authenticate the "site" as a whole (only users or posts from the site).

Uh..we're not authenticating the user or post from the site, but instead associating the token with that user or post, mostly as a storage per se, and by doing that we're also validating that the request and the site it is coming from is valid as well.

If we had the tokens persisted in a global way (let's say, saving as a option in the database), we would have the risk of making the token slightly less secure. When using the tokens from the WPJMCOM side, for instance, one token generated for each site would be valid for another site as well, which I don't think would be expected. This might be an edge-case in the case of the Promoted Jobs project, yeah, but oh well..

Also, I couldn't think an use case where it would be useful to have it persisted in a more..."global" way. Have you thought of any? I'd love to hear more of your thoughts about this.

Thanks for all these questions, by the way!

fjorgemota committed:
Handle case where base64_encode returns only invalid characters

Although rare, there's a scenario where base64_encode can return a sting composed only by the removed characters: '+', '/', '='. For instance,"+/++/+/++++/++/+++++++++/++++++///////////////////////+/+++++/+/==" is a valid Base 64 string.

Given that we there's the risk that random_bytes would return such a string (because it is, well, random). I decided to add a fallback so the string is converted to hexadecimal in cases where the base64 - after having all these characters removed - is empty.

renatho reviewed (approved):
Thank you, Fernando!

I'm approving the PR on both sides (here and on wpjobmanager.com). Are you planning to come with another PR with the requests that connect them, allowing them to create/edit the jobs on behalf of the specific site?

fjorgemota commented:
Hey there, thanks for your review!

> I'm approving the PR on both sides (here and on wpjobmanager.com). Are you planning to come with another PR with the requests that connect them, allowing them to create/edit the jobs on behalf of the specific site?

As discussed on Slack, yes. 

Thanks!
